relayer/pkg/chainlink/txm/txm.go

```diff
@@ -69,135 +81,167 @@ func (txm *starktxm) Name() string {
+	account, err := caigo.NewGatewayAccount(privKey, accountAddress.String(), client.Gw, caigo.AccountVersion1)
+	if err != nil {
+		return txhash, fmt.Errorf("failed to create new account: %+w", err)
+	}
+
+	nonce, err := txm.nonce.NextSequence(accountAddress, client.Gw.ChainId)
 	if err != nil {
-		return txhash, errors.Errorf("failed to create new account: %s", err)
+		return txhash, fmt.Errorf("failed to get nonce: %+w", err)
 	}
 
 	// get fee for txm
+	// optional - pass nonce to fee estimate (if nonce gets ahead, estimate may fail)
+	// can we estimate fee without calling estimate - tbd with 1.0
 	feeEstimate, err := account.EstimateFee(ctx, txs, caigotypes.ExecuteDetails{})
 	if err != nil {
-		return txhash, errors.Errorf("failed to estimate fee: %s", err)
+		return txhash, fmt.Errorf("failed to estimate fee: %+w", err)
 	}
 
 	fee, _ := big.NewInt(0).SetString(string(feeEstimate.OverallFee), 0)
 	expandedFee := big.NewInt(0).Mul(fee, big.NewInt(int64(FEE_MARGIN)))
 	max := big.NewInt(0).Div(expandedFee, big.NewInt(100))
 	details := caigotypes.ExecuteDetails{
 		MaxFee: max,
+		Nonce:  nonce,
 	}
 
-	// TODO: investigate if nonce management is needed (nonce is requested queried by the sdk for now)
 	// transmit txs
 	execCtx, execCancel := context.WithTimeout(ctx, txm.cfg.TxTimeout()*time.Second)
 	defer execCancel()
 	res, err := account.Execute(execCtx, txs, details)
 	if err != nil {
-		return txhash, errors.Errorf("failed to invoke tx: %s", err)
+		// TODO: handle initial broadcast errors - what kind of errors occur?
+		return txhash, fmt.Errorf("failed to invoke tx: %+w", err)
 	}
 
 	// handle nil pointer
 	if res == nil {
 		return txhash, errors.New("execute response and error are nil")
 	}
 
-	return res.TransactionHash, nil
+	// update nonce if transaction is successful
+	err = errors.Join(
+		txm.nonce.IncrementNextSequence(accountAddress, client.Gw.ChainId, nonce),
+		txm.txStore.Save(accountAddress, nonce, res.TransactionHash),
+	)
+	return res.TransactionHash, err
+}
+
+func (txm *starktxm) confirmLoop() {
+	defer txm.done.Done()
+
+	ctx, cancel := utils.ContextFromChan(txm.stop)
+	defer cancel()
+
+	tick := time.After(txm.cfg.ConfirmationPoll())
+
+	txm.lggr.Debugw("confirmLoop: started")
+	for {
+		var start time.Time
+		select {
+		case <-tick:
+			start = time.Now()
+			client, err := txm.client.Get()
+			if err != nil {
+				txm.lggr.Errorw("failed to load client", "error", err)
+				break
+			}
+
+			hashes := txm.txStore.GetUnconfirmed()
+			for addr := range hashes {
+				for i := range hashes[addr] {
+					hash := hashes[addr][i]
+					status, err := client.Gw.TransactionStatus(ctx, gateway.TransactionStatusOptions{
+						TransactionHash: hashes[addr][i],
+					})
+					if err != nil {
+						txm.lggr.Errorw("failed to fetch transaction status", "hash", hash, "error", err)
+					}
+
+					if status.TxStatus == "ACCEPTED_ON_L1" || status.TxStatus == "ACCEPTED_ON_L2" || status.TxStatus == "REJECTED" {
```

Comment:
> i don't see any code for handling rejected transactions (eg. requeueing) - does that need to be handled?

do you mean rejected by the RPC during initial broadcast or rejected by the confirmation process?
either way, we currently do not requeue - which forces the OCR2 process to trigger rebroadcasting via another CL node. also if the feed is running with a fast heartbeat a missed report can be superseded by a more recent report (like on Solana)

but I do agree that handling rejected Txs is eventually needed - this is mostly a minimal set so we are able to broadcast txs with local nonce management and understand why they may fail with the confirmation logic

> what's the point of txStore? i don't see any functional changes in the broadcast loop depending on its state. is it just so that we can calculated InflightCount?

txStore is mostly for the confirmation process so we know which txs need to be checked

Reply:
yeah in the current state - it's not a part of critical functionality, but it's really useful to see in the CL node why TXs are failing (from previous experience with Solana)

in the future, it should be replaced by the txm platform which will have retry logic built in with the confirmer